In [1]:
#Importing Libraries
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn import model_selection
from sklearn.ensemble import BaggingClassifier

C:\Users\pattn\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\ensemble\weight_boosting.py:29: DeprecationWarning: numpy.core.umath_tests is an internal NumPy module and should not be imported. It will be removed in a future NumPy release.
  from numpy.core.umath_tests import inner1d


In [2]:
#Reading data
data = pd.read_csv(r"C:/Python/Ensemble/Social_Network_Ads.csv")
data.head()

#Removing User ID, as it's an increamental value that doesn't add to our classification prediction
data.pop('User ID')

#Replacing categorical values to numericals
data['Gender'].replace(['Male','Female'],[1,0],inplace=True)

#Using features: Gender, Age for prediction of Purchased label
feature_cols = ['Gender', 'Age']
X = data[feature_cols] # Features
Y = data.Purchased # Target variable

### BAGGING

In [3]:

seed = 7
kfold = model_selection.KFold(n_splits=10, random_state=seed)
cart = DecisionTreeClassifier()
num_trees = 100
model = BaggingClassifier(base_estimator=cart, n_estimators=num_trees, random_state=seed)
results = model_selection.cross_val_score(model, X, Y, cv=kfold)
print("Accuracy for BaggingClassifier:", results.mean())

Accuracy for BaggingClassifier: 0.8


In [4]:
from sklearn.ensemble import RandomForestClassifier

seed = 7
num_trees = 100
max_features = 3
kfold = model_selection.KFold(n_splits=10, random_state=seed)
model = RandomForestClassifier(n_estimators=num_trees)
results = model_selection.cross_val_score(model, X, Y, cv=kfold)
print("Accuracy for RandomForestClassifier:", results.mean())

Accuracy for RandomForestClassifier: 0.7975


In [5]:
from sklearn.ensemble import ExtraTreesClassifier

seed = 7
num_trees = 100
max_features = 7
kfold = model_selection.KFold(n_splits=10, random_state=seed)
model = ExtraTreesClassifier(n_estimators=num_trees)
results = model_selection.cross_val_score(model, X, Y, cv=kfold)
print("Accuracy for ExtraTreesClassifier:", results.mean())

Accuracy for ExtraTreesClassifier: 0.7849999999999999


### BOOSTING

In [6]:
from sklearn.ensemble import AdaBoostClassifier

seed = 7
num_trees = 30
kfold = model_selection.KFold(n_splits=10, random_state=seed)
model = AdaBoostClassifier(n_estimators=num_trees, random_state=seed)
results = model_selection.cross_val_score(model, X, Y, cv=kfold)
print("Accuracy for AdaBoostClassifier:", results.mean())


from sklearn.ensemble import GradientBoostingClassifier

seed = 7
num_trees = 100
kfold = model_selection.KFold(n_splits=10, random_state=seed)
model = GradientBoostingClassifier(n_estimators=num_trees, random_state=seed)
results = model_selection.cross_val_score(model, X, Y, cv=kfold)
print("Accuracy for GradientBoostingClassifier:", results.mean())

Accuracy for AdaBoostClassifier: 0.835
Accuracy for GradientBoostingClassifier: 0.8149999999999998


### VOTING ENSEMBLE

In [7]:
# Voting Ensemble for Classification
from sklearn import model_selection
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.ensemble import VotingClassifier

seed = 7
kfold = model_selection.KFold(n_splits=10, random_state=seed)
# create the sub models
estimators = []
model1 = LogisticRegression()
estimators.append(('logistic', model1))
model2 = DecisionTreeClassifier()
estimators.append(('cart', model2))
model3 = SVC()
estimators.append(('svm', model3))
# create the ensemble model
ensemble = VotingClassifier(estimators)
results = model_selection.cross_val_score(ensemble, X, Y, cv=kfold)
print("Accuracy for VotingClassifier:", results.mean())

Accuracy for VotingClassifier: 0.8324999999999999


C:\Users\pattn\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\preprocessing\label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
C:\Users\pattn\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\preprocessing\label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
C:\Users\pattn\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\preprocessing\label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
C:\Users\pattn\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\preprocessing\label.py

| Classifier        | Accuracy           |
| ------------- |:-------------:|
| BaggingClassifier:       | 80 % | 
| RandomForestClassifier:      | 79.75%      |
| ExtraTreesClassifier: | 78.49 %     |
| AdaBoostClassifier:       | 83.5 % | 
| GradientBoostingClassifier:      | 81.49 %      |
| VotingClassifier: | 83.24 %     |

                                             Hence, AdaBoostClassifer performs best here.